File: examples/chapter17/e04_augment/python/main.py
"""Example 17.4 - Augmentation as a Lazy Mapping (PyTorch reference + parity test).

Mirror every image left-to-right and concatenate the mirrored copies with the
originals: 24 images become 48, and nothing is stored twice.

Run:   python main.py
Test:  pytest main.py
"""

import os

import numpy as np
import torch
from PIL import Image
from torch.utils.data import ConcatDataset, Dataset

ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "data", "shapes")


def load_all():
    items = []
    for label, c in enumerate(sorted(os.listdir(ROOT))):
        for f in sorted(os.listdir(os.path.join(ROOT, c))):
            pixels = torch.from_numpy(np.array(Image.open(os.path.join(ROOT, c, f)).convert("L")))
            items.append((pixels, label))
    return items


class Mapped(Dataset):
    def __init__(self, items, fn):
        self.items, self.fn = items, fn

    def __len__(self):
        return len(self.items)

    def __getitem__(self, i):
        pixels, label = self.items[i]
        return self.fn(pixels), label   # applied on every access, like a Burn Mapper


def mirror(x):
    return torch.flip(x, dims=[-1])


def build():
    base = load_all()
    augmented = ConcatDataset([Mapped(base, lambda x: x), Mapped(base, mirror)])
    original, label0 = augmented[0]
    mirrored, label1 = augmented[len(base)]
    bright = int(original.int().sum(dim=1).argmax())
    round_trip = torch.equal(mirror(mirrored), original) and label0 == label1
    return (len(base), len(augmented), original[bright].tolist(), mirrored[bright].tolist(),
            round_trip)


def main():
    n, n_aug, before, after, round_trip = build()
    print(f"dataset: {n} images -> with mirrored copies: {n_aug}")
    print("image 0, bright row          =", before)
    print(f"image {n}, same row, mirrored =", after)
    print("mirror twice == original, label kept:", round_trip)


def test_matches_burn():
    n, n_aug, before, after, round_trip = build()
    assert (n, n_aug) == (24, 48)
    assert after == before[::-1]
    assert before == EXPECTED_BEFORE
    assert round_trip


EXPECTED_BEFORE = [24, 32, 3, 18, 22, 231, 229, 239, 223, 251, 245, 17]


if __name__ == "__main__":
    main()

File: examples/chapter17/e04_augment/rust/src/main.rs
use std::sync::Arc;

use burn::data::dataset::transform::{ComposedDataset, Mapper, MapperDataset};
use burn::data::dataset::{Dataset, InMemDataset};

const ROOT: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../data/shapes"
);
const SIZE: usize = 12;

#[derive(Clone, Debug)]
struct ImageItem {
    pixels: Vec<u8>, // 12 x 12, row by row
    label: usize,
}

// Example 17.2 does this lazily; 24 tiny images can simply be read up front.
fn load_all() -> InMemDataset<ImageItem> {
    let mut classes: Vec<_> = std::fs::read_dir(ROOT)
        .unwrap()
        .map(|e| e.unwrap().path())
        .collect();
    classes.sort();
    let mut items = Vec::new();
    for (label, dir) in classes.iter().enumerate() {
        let mut files: Vec<_> = std::fs::read_dir(dir)
            .unwrap()
            .map(|e| e.unwrap().path())
            .collect();
        files.sort();
        for f in files {
            let pixels = image::open(f).unwrap().to_luma8().into_raw();
            items.push(ImageItem { pixels, label });
        }
    }
    InMemDataset::new(items)
}

// A Mapper turns one item into another. MapperDataset applies it inside get(),
// so nothing is precomputed and nothing is stored twice.
#[derive(Clone, Copy)]
enum Augment {
    Identity,
    MirrorLeftRight,
}

impl Mapper<ImageItem, ImageItem> for Augment {
    fn map(&self, item: &ImageItem) -> ImageItem {
        match self {
            Augment::Identity => item.clone(),
            Augment::MirrorLeftRight => {
                let mut pixels = Vec::with_capacity(SIZE * SIZE);
                for row in item.pixels.chunks(SIZE) {
                    pixels.extend(row.iter().rev());
                }
                ImageItem {
                    pixels,
                    label: item.label,
                } // the label does not change
            }
        }
    }
}

fn row(item: &ImageItem, r: usize) -> Vec<u8> {
    item.pixels[r * SIZE..(r + 1) * SIZE].to_vec()
}

/// Returns (original len, augmented len, the bright row of image 0, the same row mirrored, round trip ok).
fn build() -> (usize, usize, Vec<u8>, Vec<u8>, bool) {
    let base = Arc::new(load_all());

    // Same Dataset TYPE for both halves (MapperDataset<_, Augment, _>), so they can be composed.
    let augmented = ComposedDataset::new(vec![
        MapperDataset::new(base.clone(), Augment::Identity),
        MapperDataset::new(base.clone(), Augment::MirrorLeftRight),
    ]);

    let original = augmented.get(0).unwrap(); // first half: untouched
    let mirrored = augmented.get(base.len()).unwrap(); // second half: the same image, mirrored

    // Image 0 is a horizontal bar: find its brightest row.
    let bright = (0..SIZE)
        .max_by_key(|&r| row(&original, r).iter().map(|&p| p as u32).sum::<u32>())
        .unwrap();
    let back = Augment::MirrorLeftRight.map(&mirrored);

    (
        base.len(),
        augmented.len(),
        row(&original, bright),
        row(&mirrored, bright),
        back.pixels == original.pixels && mirrored.label == original.label,
    )
}

fn main() {
    let (n, n_aug, before, after, round_trip) = build();
    println!("dataset: {n} images -> with mirrored copies: {n_aug}");
    println!("image 0, bright row          = {before:?}");
    println!("image {n}, same row, mirrored = {after:?}");
    println!("mirror twice == original, label kept: {round_trip}");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matches_pytorch() {
        let (n, n_aug, before, after, round_trip) = build();
        assert_eq!((n, n_aug), (24, 48));
        let expected: Vec<u8> = before.iter().rev().copied().collect(); // torch.flip(x, [-1])
        assert_eq!(after, expected);
        assert_eq!(
            before,
            vec![24, 32, 3, 18, 22, 231, 229, 239, 223, 251, 245, 17]
        );
        assert!(round_trip);
    }
}
